File: 53-maximum-subarray/53-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        @cache
        def helper(pos: int, pick: bool) -> int:
            if pos == n:
                return 0 if pick else -inf
            
            if pick:
                return max(
                    nums[pos] + helper(pos + 1, True),
                    0
                )
            else:
                return max(
                    nums[pos] + helper(pos + 1, True),
                    helper(pos + 1, False)
                )
        n = len(nums)
        return helper(0, False)

File: 53-maximum-subarray/53-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int currPicked = 0, currNotPicked = Integer.MIN_VALUE;
        for(int i = nums.length - 1; i >= 0; i--){
            int prevPicked = currPicked;
            currPicked = Math.max(
                nums[i] + prevPicked,
                0
            );
            currNotPicked = Math.max(
                nums[i] + prevPicked,
                currNotPicked
            );
        }
        return currNotPicked;
    }
}